### ISSUE TITLE
NextFlowPlatform 2.0

### ISSUE THREAD

**Issue by rcannood** [viash-io/viash#82]
I created a repository, [`viash_nxf_poc`](https://github.com/viash-io/viash_nxf_poc), to work out a POC for a NextFlowPlatform rewrite. I'm proposing this rewrite to fix some of my annoyances with the current way of working, but also to reduce the code complexity and add more checks in order to avoid bugs (which currently occur quite regularly).

---

### Channel Interface

A Viash+Nextflow module generated by Viash has the interface:

```
Input channel:  [id, inputs, ...passthrough...]
Output channel: [id, outputs, ...passthrough...]
```

These fields are defined as follows:
* `id` (`String`) is a unique identifier for the event in the Channel.
* `inputs` (`Map[String, Object]` or `File`) is a named map containing the component's input parameters.
  Examples of the class types associated with different Viash component arguments:
  
  | config.vsh.yaml | Input map in Nextflow | Class |
  | ------- | -------- | --- |
  | `{ name: foo, type: string, direction: input }` | `[ foo: "bar" ]` | `String` |
  | `{ name: int, type: integer, direction: input, multiple: true }` | `[ int: [ 1, 2, 3 ] ]` | `List[Integer]` |
  | `{ name: bool, type: boolean, direction: input, required: false }` | `[ bool: null ]` | `Boolean` or `null` |
  | `{ name: bool, type: boolean, direction: input, required: false }` | `[ bool: true ]` | `Boolean` or `null` |
  | `{ name: in, type: file, direction: input }` | `[ in: file("in.h5ad") ]` | `File` |
  | `{ name: out, type: file, direction: output }` | `[ out: "proposed_path.h5ad" ]` | `String` |
  | `{ name: out, type: file, direction: output, multiple: true }` | `[ out: "proposed_path_*.h5ad" ]` | `String` |
  
  If you only want to specify a single input file, you can simply pass a `File` instead of a `Map[String, Object]`.
* `...passthrough...` (`Object*`) are objects that simply get passed through to the output. This is a practical solution for reading a bunch of parameters in at the start of a workflow and putting it into the inputs slot whenever they need to be consumed. This means that an event in the channel can be of length N where N >= 2.
* `outputs` (`Map[String, File]` or `File`) is a named map containing the component's output files. If the component outputs only a single File, the outputs will be a File rather than a named map.

---

### Module usage

Given a Viash component named `poc` ( `src/poc/config.vsh.yaml` ), importing the module yields a Nextflow Workflow which can be used as follows:

```groovy
nextflow.enable.dsl=2

include { poc } from "./target/nextflow/poc/main.nf" params(params)

workflow {
  Channel.value( [
    "foo", 
    [
      input_one: file("data/pbmc_1k_protein_v3.normalize.output_rna.h5ad"),
      input_multi: file("data/*.h5ad"),
      string: "foo",
      integer: 10
    ]
  ])
  | poc
}
```

---

### Viash+Nextflow modules are flexible

The strength of the new Viash+Nextflow modules lies in its flexibility in how you want to use the module.

```groovy
nextflow.enable.dsl=2

include { poc } from "./target/nextflow/poc/main.nf"

workflow {
  Channel.value( [
    "foo", 
    [
      input_one: file("data/pbmc_1k_protein_v3.normalize.output_rna.h5ad"),
      input_multi: file("data/*.h5ad")
    ]
  ])
  | poc.run(
    args: [ string: "foo", integer: 10 ],
    directives: [
      cache: "lenient",
      label: [ "bigmem", "bigcpu" ]
    ],
    auto: [
      simplifyInput: true,
      simplifyOutput: true,
      publish: false,
      transcript: false
    ]
  )
}
```

* `directives`: One on one mapping with the [Nextflow process directives](https://www.nextflow.io/docs/latest/process.html#directives). **NOTE:** You _can_ pass clojures, but they need to quoted, see example below. Examples:
    - `container: "bash:4.2"`
    - `label: ["bigmem", "bigcpu"]`
    - `publishDir: [ path: "output/", mode: "copy", saveAs: "{ "prefix_" + it }" ]` (← saveAs is a quoted closure)
* `auto`: Helper arguments provided by Viash.
  - `simplifyInput`: If `true`, an input tuple only containing only a single File (e.g. `["foo", file("in.h5ad")]`) is automatically transformed to a map (i.e. `["foo", [ input: file("in.h5ad") ] ]`)
  - `simplifyOutput`: If `true`, an output tuple containing a map with a File (e.g. `["foo", [ output: file("out.h5ad") ] ]`) is automatically transformed to a map (i.e. `["foo", file("out.h5ad")]`)
  - `publish`: If `true`, the module's outputs are automatically published to `params.publishDir`. Will throw an error if `params.publishDir` is not defined.
  - `transcript`: If `true`, the module's transcripts are automatically published to `params.transcriptDir`. If not defined, `params.publishDir + "/_transcripts"` will be used. Will throw an error if neither are defined.

---

### Chaining multiple modules

If each module only has one input file and output file:

```groovy
nextflow.enable.dsl=2

include { poc1 } from "./target/nextflow/poc1/main.nf"
include { poc2 } from "./target/nextflow/poc2/main.nf"
include { poc3 } from "./target/nextflow/poc3/main.nf"

workflow {
  Channel.value( [ "foo", file("data/pbmc_1k_protein_v3.normalize.output_rna.h5ad") ] )
  | poc1
  | poc2
  | poc3
```

If the modules have multiple input / output files per step:
```groovy
nextflow.enable.dsl=2

include { poc1 } from "./target/nextflow/poc1/main.nf"
include { poc2 } from "./target/nextflow/poc2/main.nf"
include { poc3 } from "./target/nextflow/poc3/main.nf"

workflow {
  Channel.value( [
    "foo", 
    [
      input_one: file("data/pbmc_1k_protein_v3.normalize.output_rna.h5ad"),
      input_multi: file("data/*.h5ad")
    ]
  ])
  | poc1.run(
    args: [ string: "foo", integer: 10 ]
  )
  | poc2.run(
    renameKeys: ["input_one": "output_one", "input_multi": "output_multi"]
  )
  | poc3.run(
    mapData: { [input_one: it.output_one, input_multi: it.output_multi ] }
  )
}
```

* `map`: Apply a map over the incoming tuple. 
   Example: `{ tup -> [ tup[0], [input: tup[1].output], tup[2] ] }`.
* `mapId`: Apply a map over the ID element of a tuple (i.e. the first element). 
   Example: `{ id -> id + "_foo" }`
* `mapData`: Apply a map over the data element of a tuple (i.e. the second element).
   Example: `{ data -> [ input: data.output ] }`
* `mapPassthrough`: Apply a map over the passthrough elements of a tuple (i.e. the tuple excl. the first two elements). 
   Example: `{ pt -> pt.drop(1) }`
* `renameKeys`: Rename keys in the data field of the tuple (i.e. the second element).
   Example: `[ "new_key": "old_key" ]`
* `debug`: Whether or not to print debug messages.
   Example: `true`

---

### Reuse same module

You can run the same component multiple times. For reasons, you need to specify a unique key every time the module is used.

```groovy
nextflow.enable.dsl=2

include { poc } from "./target/nextflow/poc/main.nf"

workflow {
  Channel.value( [ "foo", file("data/pbmc_1k_protein_v3.normalize.output_rna.h5ad") ] )
  | poc.run(key: "step1")
  | poc.run(key: "step2")
  | poc.run(key: "step3")
```

**Comment by rcannood** [comment]
Going over all directives to determine how they should be managed.

### [accelerator](https://www.nextflow.io/docs/latest/process.html#accelerator)
Config:
```yaml
platforms:
  - type: nextflow
    directives:
      accelerator: "4, type: 'nvidia-tesla-k80'"
```

Or in the function call:
```groovy
foo(
  directives: [
    "accelerator": "4, type: 'nvidia-tesla-k80'"
  ]
)
```

Becomes
```nextflow
process foo {
  accelerator 4, type: 'nvidia-tesla-k80'
}
```

**Comment by rcannood** [comment]
@tverbeiren Did I forget something? I'm going to use the content of this issue to make a blog post on viash.io.

**Comment by rcannood** [comment]
This functionality was released in 0.5.11 :partying_face:
